File: vendor_portal/vendor_portal/page/vendor_dashboard/vendor_dashboard.js
frappe.pages["vendor-dashboard"].on_page_load = function (wrapper) {
	frappe.ui.make_app_page({
		parent: wrapper,
		title: "Vendor Dashboard",
		single_column: true,
	});

	let page = wrapper.page;
	let $body = $(wrapper).find(".page-content");

	// Quick navigation to Comparison Matrix
	page.add_menu_item(__("Vendor Comparison Matrix"), () => frappe.set_route("vendor-comparison"));

	$body.html(`
		<div class="vendor-dashboard">
			<!-- KPI Strip -->
			<div class="kpi-strip">
				<div class="kpi-card" id="kpi-suppliers">
					<div class="kpi-label">${__("Active Suppliers")}</div>
					<div class="kpi-value">—</div>
				</div>
				<div class="kpi-card" id="kpi-onboardings">
					<div class="kpi-label">${__("Pending Onboardings")}</div>
					<div class="kpi-value">—</div>
				</div>
				<div class="kpi-card" id="kpi-pos">
					<div class="kpi-label">${__("POs This Month")}</div>
					<div class="kpi-value">—</div>
				</div>
				<div class="kpi-card" id="kpi-avg-rating">
					<div class="kpi-label">${__("Avg Scorecard Score")}</div>
					<div class="kpi-value">—</div>
				</div>
				<div class="kpi-card" id="kpi-compare" style="cursor:pointer;border:1px dashed var(--border-color);background:transparent" onclick="frappe.set_route('vendor-comparison')">
					<div class="kpi-label">${__("Vendor Comparison")}</div>
					<div class="kpi-value" style="font-size:20px">⚡ <span style="font-size:13px;font-weight:600;color:#3b82f6">${__("Compare →")}</span></div>
				</div>
			</div>

			<!-- Chart Grid -->
			<div class="chart-grid">
				<div class="chart-card wide">
					<div class="chart-header">
						<h4>${__("Vendor Rating Distribution")}</h4>
						<span class="chart-subtitle">${__("Supplier Scorecard score buckets")}</span>
					</div>
					<div class="chart-wrapper" id="chart-rating-distribution"></div>
				</div>
				<div class="chart-card">
					<div class="chart-header">
						<h4>${__("Onboarding Pipeline")}</h4>
						<span class="chart-subtitle">${__("Applications by status")}</span>
					</div>
					<div class="chart-wrapper" id="chart-onboarding-pipeline"></div>
				</div>
				<div class="chart-card">
					<div class="chart-header">
						<h4>${__("PO Volume by Category")}</h4>
						<span class="chart-subtitle">${__("Total purchase value per vendor category")}</span>
					</div>
					<div class="chart-wrapper" id="chart-po-volume"></div>
				</div>
				<div class="chart-card wide">
					<div class="chart-header">
						<h4>${__("Delivery Performance Trend")}</h4>
						<span class="chart-subtitle">${__("Average monthly scorecard score (last 12 months)")}</span>
					</div>
					<div class="chart-wrapper" id="chart-delivery-trend"></div>
				</div>
			</div>
		</div>
	`);

	// Load KPI counts
	_load_kpis();

	// Load all 4 charts
	_load_chart_rating_distribution();
	_load_chart_onboarding_pipeline();
	_load_chart_po_volume();
	_load_chart_delivery_trend();
};

// ─────────────────────────────────────────────────────────────────────────
// KPI helpers
// ─────────────────────────────────────────────────────────────────────────

function _load_kpis() {
	Promise.all([
		frappe.db.count("Supplier", { disabled: 0 }),
		frappe.db.count("Vendor Onboarding", { onboarding_status: "Under Review" }),
		frappe.xcall("frappe.client.get_list", {
			doctype: "Purchase Order",
			filters: [
				["docstatus", "=", 1],
				["transaction_date", ">=", frappe.datetime.month_start()],
			],
			fields: ["name"],
			limit_page_length: 0,
		}),
		frappe.xcall("frappe.client.get_list", {
			doctype: "Supplier Scorecard",
			fields: ["supplier_score"],
			limit_page_length: 0,
		}),
	]).then(([suppliers, onboardings, pos, scorecards]) => {
		$("#kpi-suppliers .kpi-value").text(suppliers);
		$("#kpi-onboardings .kpi-value").text(onboardings);
		$("#kpi-pos .kpi-value").text(pos.length);

		let avg = 0;
		if (scorecards && scorecards.length) {
			let total = scorecards.reduce((s, r) => s + flt(r.supplier_score), 0);
			avg = (total / scorecards.length).toFixed(1);
		}
		$("#kpi-avg-rating .kpi-value").text(avg || "—");
	});
}

// ─────────────────────────────────────────────────────────────────────────
// Chart loaders
// ─────────────────────────────────────────────────────────────────────────

function _load_chart_rating_distribution() {
	frappe
		.call("vendor_portal.vendor_portal.page.vendor_dashboard.vendor_dashboard.get_rating_distribution")
		.then((r) => {
			if (!r.message) return;
			new frappe.Chart("#chart-rating-distribution", {
				data: r.message,
				type: "bar",
				height: 280,
				colors: ["#2490ef"],
				barOptions: { spaceRatio: 0.6 },
				axisOptions: { xAxisMode: "tick", yAxisMode: "span" },
			});
		});
}

function _load_chart_onboarding_pipeline() {
	frappe
		.call("vendor_portal.vendor_portal.page.vendor_dashboard.vendor_dashboard.get_onboarding_pipeline")
		.then((r) => {
			if (!r.message) return;
			new frappe.Chart("#chart-onboarding-pipeline", {
				data: r.message,
				type: "bar",
				height: 280,
				colors: ["#7c3aed"],
				barOptions: { spaceRatio: 0.5 },
				axisOptions: { xAxisMode: "tick" },
			});
		});
}

function _load_chart_po_volume() {
	frappe
		.call("vendor_portal.vendor_portal.page.vendor_dashboard.vendor_dashboard.get_po_volume_by_category")
		.then((r) => {
			if (!r.message) return;
			new frappe.Chart("#chart-po-volume", {
				data: r.message,
				type: "bar",
				height: 280,
				colors: ["#d97706"],
				barOptions: { spaceRatio: 0.5 },
				axisOptions: { xAxisMode: "tick" },
			});
		});
}

function _load_chart_delivery_trend() {
	frappe
		.call("vendor_portal.vendor_portal.page.vendor_dashboard.vendor_dashboard.get_delivery_trend")
		.then((r) => {
			if (!r.message) return;
			new frappe.Chart("#chart-delivery-trend", {
				data: r.message,
				type: "line",
				height: 280,
				colors: ["#059669"],
				lineOptions: { regionFill: 1, hideDots: 0, heatLine: 1 },
				axisOptions: { xIsSeries: 1, xAxisMode: "tick", yAxisMode: "span" },
			});
		});
}

File: vendor_portal/vendor_portal/page/vendor_dashboard/vendor_dashboard.py
# Copyright (c) 2026, renish and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import flt


@frappe.whitelist()
def get_rating_distribution():
	"""Returns histogram data: count of suppliers in each scorecard bucket."""
	buckets = [
		{"min": 0, "max": 20, "label": "0-20 (Critical)"},
		{"min": 20, "max": 40, "label": "20-40 (Very Poor)"},
		{"min": 40, "max": 60, "label": "40-60 (Poor)"},
		{"min": 60, "max": 80, "label": "60-80 (Average)"},
		{"min": 80, "max": 101, "label": "80-100 (Excellent)"},
	]

	rows = frappe.db.sql(
		"""
		SELECT sc.supplier_score
		FROM `tabSupplier Scorecard` sc
		JOIN `tabSupplier` s ON s.name = sc.supplier
		WHERE s.disabled = 0
	""",
		as_dict=1,
	)

	scores = [flt(r.supplier_score) for r in rows if r.supplier_score is not None]

	labels = []
	counts = []
	for b in buckets:
		c = sum(1 for s in scores if b["min"] <= s < b["max"])
		labels.append(b["label"])
		counts.append(c)

	return {"labels": labels, "datasets": [{"name": "Suppliers", "values": counts}]}


@frappe.whitelist()
def get_onboarding_pipeline():
	"""Returns count of Vendor Onboardings grouped by status."""
	rows = frappe.db.sql(
		"""
		SELECT onboarding_status, COUNT(*) AS count
		FROM `tabVendor Onboarding`
		GROUP BY onboarding_status
		ORDER BY FIELD(onboarding_status, 'Draft', 'Under Review', 'Approved', 'Rejected')
	""",
		as_dict=1,
	)

	labels = [r.onboarding_status for r in rows]
	counts = [r.count for r in rows]

	return {"labels": labels, "datasets": [{"name": "Onboardings", "values": counts}]}


@frappe.whitelist()
def get_po_volume_by_category():
	"""Returns total PO value per Vendor Category."""
	rows = frappe.db.sql(
		"""
		SELECT
			COALESCE(s.custom_vendor_category, 'Uncategorized') AS category,
			SUM(po.base_grand_total) AS total_value
		FROM `tabPurchase Order` po
		JOIN `tabSupplier` s ON s.name = po.supplier
		WHERE po.docstatus = 1
		GROUP BY s.custom_vendor_category
		ORDER BY total_value DESC
	""",
		as_dict=1,
	)

	labels = [r.category for r in rows]
	values = [flt(r.total_value) for r in rows]

	return {"labels": labels, "datasets": [{"name": "PO Value", "values": values}]}


@frappe.whitelist()
def get_delivery_trend():
	"""Returns monthly average Supplier Scorecard Period score for the last 12 months."""
	rows = frappe.db.sql(
		"""
		SELECT
			DATE_FORMAT(scp.end_date, '%Y-%m') AS month,
			AVG(scp.total_score) AS avg_score
		FROM `tabSupplier Scorecard Period` scp
		WHERE scp.docstatus = 1
			AND scp.end_date >= DATE_SUB(CURDATE(), INTERVAL 12 MONTH)
		GROUP BY month
		ORDER BY month ASC
	""",
		as_dict=1,
	)

	labels = [r.month for r in rows]
	scores = [flt(r.avg_score) for r in rows]

	return {"labels": labels, "datasets": [{"name": "Avg Score", "values": scores}]}
